### PR TITLE
stmhal: SPI Todos 

### PR DESCRIPTION
**Pull request by hoihu**
tried to resolve some todos from the spi module:
- added missing documentation of already implemented options (nss, dir)
- added constants for direction (dir) and slave select (nss) options
- updatet spi unit tests
- printout of dir and nss options

adds 348 bytes - mainly due to the printout (may be simplified or optimised)

### CODE REVIEW

**Comment by dpgeorge** [comment]
The reason it was a TODO is because it was unclear how to proceed.  But I guess you did the simplest thing by just providing the constants and making the print print everything.

The constants don't need to follow the ST names, and in fact I think they should be simplified.  Eg it's enough to just have SOFT, HARD_IN and HARD_OUT for the nss constants.

The print function can definitely be optimised.  Eg printing qstrs instead of C strings.

Note that docs now live in docs/library/pyb.SPI.rst and the need updating.  The inline comment-documentation (starting with ///) is now obsolete.

File: stmhal/spi.c [hunk 1]
```diff
@@ -747,18 +786,22 @@ STATIC const mp_map_elem_t pyb_spi_locals_dict_table[] = {
...
-    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_SOFT                    SPI_CR1_SSM
-    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_HARD_INPUT              ((uint32_t)0x00000000)
-    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_HARD_OUTPUT             ((uint32_t)0x00040000)
-    */
+    { MP_OBJ_NEW_QSTR(MP_QSTR_DIRECTION_2LINES), MP_OBJ_NEW_SMALL_INT(SPI_DIRECTION_2LINES) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_DIRECTION_2LINES_RXONLY), MP_OBJ_NEW_SMALL_INT(SPI_CR1_RXONLY) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_DIRECTION_1LINE), MP_OBJ_NEW_SMALL_INT(SPI_CR1_BIDIMODE) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_SOFT),        MP_OBJ_NEW_SMALL_INT(SPI_CR1_SSM) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_HARD_INPUT),  MP_OBJ_NEW_SMALL_INT(SPI_NSS_HARD_INPUT) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_HARD_OUTPUT), MP_OBJ_NEW_SMALL_INT(SPI_NSS_HARD_INPUT) }
```
Comment (dpgeorge) [inline stmhal/spi.c]: The constant here is SPI_NSS_HARD_INPUT, should it be OUTPUT?

Comment (hoihu) [inline stmhal/spi.c]: Yes you're right.

**Comment by hoihu** [comment]
> The inline comment-documentation (starting with ///) is now obsolete.

should the inline comments be removed then and a link placed to the rst file? 

A mix between the two might also be an option: keep the inline comments moreless up-to-date with added informations for programmers and a shorter version for the rst file.

> The constants don't need to follow the ST names, and in fact I think they should be simplified
 
DIRECTION could be cut short as well (e.g. DIR_1L, DIR_2L, etc.)

**Comment by hoihu** [comment]
After reading through the datasheets, a proposal for a clearer, yet shorter version of the constants:

1) Use `wires` instead of `dir`:
`2WIRE` (standard MISO, MOSI)
`1WIRE_RX` (receive only)
`1WIRE_BI` (bidrectional)
2) Use `cs`instead of `nss`: This makes it more "pythonic" for HW developer as I believe you'd commonly use chip select signals, not slave select. I propose the following options:
`CS_SW` (chip select using fw)
`CS_HW_OUT`(chip select using pin output - for masters)
`CS_HW_IN` (chip select using pin input - for slaves)

Note that for 2) the meaning inverses (SPI_NSS_HARD_INPUT actually is CS_HW_OUT).

@dpgeorge Is this a way to go? I'm happy to update the RST as well, let me know what to do with the code documentation (if both should be updated).

File: stmhal/spi.c [hunk 2]
```diff
@@ -747,18 +784,22 @@ STATIC const mp_map_elem_t pyb_spi_locals_dict_table[] = {
...
-    { MP_OBJ_NEW_QSTR(MP_QSTR_DIRECTION_2LINES             ((uint32_t)0x00000000)
-    { MP_OBJ_NEW_QSTR(MP_QSTR_DIRECTION_2LINES_RXONLY      SPI_CR1_RXONLY
-    { MP_OBJ_NEW_QSTR(MP_QSTR_DIRECTION_1LINE              SPI_CR1_BIDIMODE
-    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_SOFT                    SPI_CR1_SSM
-    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_HARD_INPUT              ((uint32_t)0x00000000)
-    { MP_OBJ_NEW_QSTR(MP_QSTR_NSS_HARD_OUTPUT             ((uint32_t)0x00040000)
-    */
+    { MP_OBJ_NEW_QSTR(MP_QSTR_TWO_WIRE), MP_OBJ_NEW_SMALL_INT(SPI_DIRECTION_2LINES) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_ONE_WIRE_RX), MP_OBJ_NEW_SMALL_INT(SPI_CR1_RXONLY) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_ONE_WIRE_BI), MP_OBJ_NEW_SMALL_INT(SPI_CR1_BIDIMODE) },
```
Comment (tobbad) [inline stmhal/spi.c]: These two macors are handled in HAL. Better introduce SPI_DIRECTION_1LINE. 
For a working implementation see #2058.

**Comment by tobbad** [comment]
Python NAMES are not allowed to start with a number:
https://docs.python.org/3/reference/lexical_analysis.html
See my note on #2058

**Comment by hoihu** [comment]
Fair enough. Something like ONE_WIRE_RX etc might do. 

I guess the difficult thing is to find a balance between long definition names which are easier to understand (but occupy more space..). And shorter "micro" like definitions.

**Comment by dpgeorge** [comment]
I'll close this due to inactivity.

The duplex/one-wire option is a nice addition so fell free to rework that on top of the latest master as a change to `machine.SPI` (see `machine_spi.c`) and open another PR if you like (would also be good to make the API more general so it can be implemented on other ports).

**Pull request closed without merging**